snapcraft/plugins/catkin_tools.py

```diff
@@ -0,0 +1,124 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The catkin_tools build plugin is useful for building ROS parts.
+
+This plugin relies on the catkin plugin as the catkin plugin performs
+configuration that allows catkin_tools to run.  This plugin uses the
+same keywords and configurations as the catkin plugin, the difference
+is the installation of and using catkin_tools to build.
+"""
+
+import os
+import logging
+
+import snapcraft.plugins.catkin
+from snapcraft.plugins.catkin import (
+    Compilers,
+)
+
+import snapcraft
+
+logger = logging.getLogger(__name__)
+
+
+class CatkinToolsPlugin(snapcraft.plugins.catkin.CatkinPlugin):
+    def __init__(self, name, options, project):
+        super().__init__(name, options, project)
+        self.build_packages.extend(['python-catkin-tools'])
+
+    def _prepare_build(self):
+        super()._prepare_build()
+
+        self._clean_catkin()
+
+        self._add_catkin_profile()
+
+        self._configure_catkin_profile()
+
+    def _clean_catkin(self):
+        # Clean to prevent conflicts with externally built packages.
+        catkincmd = ['catkin', 'clean', '-y']
+
+        self._run_in_bash(catkincmd)
+
+    def _add_catkin_profile(self):
+        # Overwrite the default catkin profile to ensure builds
+        # aren't affected by profile changes.
+        catkincmd = ['catkin', 'profile', 'add', '-f', 'default']
```

Comment:
Is there any downside to potentially overwriting this profile? Is it possible for the developer to create one with settings that are necessary for the build to succeed, and our overwriting it here causes it to fail? Likewise, I'm sure it's possible for their changes to break the build in the snap... perhaps that's more likely. Just want to get your thoughts on this.

Reply:
As a user I would probably prefer to be able to use in my own profile.  It would mean the catkin tools plugin would be less maintenance as well, since almost anything you would want to add can be done directly through a profile.  My thought was that a user's settings could break the build in the snap if they were utilizing some features, be using source not in the parts directory, or not install properly.  Would a field in the schema be appropriate to allow the user to use their schema understanding it could break things?  Even if we allow the user's profile we would still want to change the source, build, and install space.

Reply:
Your logic seems sound, here. Indeed, the schema seems a good place for this, but I think what we have here is a good and safe first step. Anything more could come as a follow-up.